hail/src/main/scala/is/hail/expr/ir/PValue.scala

```diff
@@ -1,27 +1,92 @@
 package is.hail.expr.ir
 
-import is.hail.asm4s.{Code, TypeInfo}
-import is.hail.expr.types.physical.{PType, PVoid}
+import is.hail.annotations.{Region, UnsafeUtils}
+import is.hail.asm4s._
+import is.hail.expr.types.physical.{PCanonicalArray, PCanonicalDict, PCanonicalSet, PContainer, PType, PVoid}
 
-abstract class PSettable {
-  def load(): PValue
+abstract class PSettable[PV <: PValue] {
+  def load(): PV
 
-  def store(v: PValue): Code[Unit]
+  def store(v: PV): Code[Unit]
 
-  def :=(v: PValue): Code[Unit] = store(v)
+  def :=(v: PV): Code[Unit] = store(v)
+
+  def storeAny(v: PValue): Code[Unit] = store(v.asInstanceOf[PV])
 }
 
 object PValue {
+  def apply(pt: PType, code: Code[_]): PValue = pt match {
+    case pt: PCanonicalArray =>
+      new PCanonicalIndexableValue(pt, coerce[Long](code))
+    case pt: PCanonicalSet =>
+      new PCanonicalIndexableValue(pt, coerce[Long](code))
+    case pt: PCanonicalDict =>
+      new PCanonicalIndexableValue(pt, coerce[Long](code))
+
+    case _ =>
+      new PrimitivePValue(pt, code)
+  }
+
   def _empty: PValue = PValue(PVoid, Code._empty)
 }
 
-case class PValue(
-  pt: PType,
-  code: Code[_]) {
+abstract class PValue {
+  def pt: PType
+
+  def code: Code[_]
+
+  def typeInfo: TypeInfo[_] = typeToTypeInfo(pt)
+
   def tcode[T](implicit ti: TypeInfo[T]): Code[T] = {
     assert(ti == typeInfo)
     code.asInstanceOf[Code[T]]
   }
+}
 
-  def typeInfo: TypeInfo[_] = typeToTypeInfo(pt)
+class PrimitivePValue(val pt: PType, val code: Code[_]) extends PValue
+
+abstract class PIndexableValue extends PValue {
+  def loadLength(): Code[Int]
+
+  def isElementDefined(i: Code[Int]): Code[Boolean]
+
+  def loadElement(length: Code[Int], i: Code[Int]): PValue
+
+  def loadElement(i: Code[Int]): PValue = loadElement(loadLength(), i)
+
+  def isElementMissing(i: Code[Int]): Code[Boolean] = !isElementDefined(i)
+}
+
+class PCanonicalIndexableValue(val pt: PContainer, val a: Code[Long]) extends PIndexableValue {
+  def code: Code[_] = a
+
+  def elementType: PType = pt.elementType
+
+  def arrayElementSize: Long = UnsafeUtils.arrayElementSize(elementType)
+
+  def loadLength(): Code[Int] = Region.loadInt(a)
+
+  def nMissingBytes(len: Code[Int]): Code[Int] = (len + 7) >>> 3
```

Comment:
this could use UnsafeUtils.packBitsToBytes

Reply:
I'm very much against the name packBitsToBytes.  Not only is it actually longer than the expression above, the above expression is also clearer in its intent and based on the name the meaning is completely opaque.  I have no idea what "pack" means, but that function certainly does not pack.  I will leave it as is.

Reply:
The name was meant to refer to bit packing: we take n bits and store them as n/8 bytes; we use this many places so I thought it reasonable to make a function for. For my education: what would you call this (or just leave out entirely)? nMissingBytes?

Reply:
I think I originally called it nMissingBytes, yes, which now calls packBitsToBytes

 >   def nMissingBytes(len: Code[Int]): Code[Int] = UnsafeUtils.packBitsToBytes(len)

I wouldn't call the expression `(n + 7) >> 3` anything.

Reply:
Got it, thanks